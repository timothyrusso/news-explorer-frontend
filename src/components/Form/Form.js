import React, { useRef } from 'react';
import { useDispatch } from 'react-redux';
import {
  setIsFormValidityTrueAction,
  setIsFormValidityFalseAction,
} from '../../store/errors/errors.actions';
import './Form.css';

const Form = ({ name, onSubmit, children }) => {
  const formRef = useRef();
  const dispatch = useDispatch();

  const onFormUpdate = (data) => {
    data
      ? dispatch(setIsFormValidityTrueAction())
      : dispatch(setIsFormValidityFalseAction());
  };

  const handleFormValidity = () => {
    const formValidityBoolean = formRef.current.checkValidity();
    onFormUpdate(formValidityBoolean);
  };

  return (
    <form
      action="#"
      className="popup__form"
      name={`myForm${name}`}
      onSubmit={onSubmit}
      ref={formRef}
      onChange={handleFormValidity}
    >
      {children}
    </form>
  );
};

export default Form;
