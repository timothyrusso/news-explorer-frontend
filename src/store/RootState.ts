import { UserState } from './user/user.reducer';
import { ArticleState } from './article/article.reducer';
import { TogglesState } from './toggles/toggles.reducer';
import { ErrorsState } from './errors/errors.reducer';

export type RootState = {
  user: UserState;
  article: ArticleState;
  toggles: TogglesState;
  errors: ErrorsState;
};
