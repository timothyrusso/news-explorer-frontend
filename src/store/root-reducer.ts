import { combineReducers } from 'redux';
import { userReducer } from './user/user.reducer';
import { articleReducer } from './article/article.reducer';

export const rootReducer = combineReducers({
  user: userReducer,
  article: articleReducer,
});
