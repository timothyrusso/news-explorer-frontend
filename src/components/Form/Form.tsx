import React, { useRef, FC } from 'react';
import { useDispatch } from 'react-redux';
import {
  setIsFormValidityTrueAction,
  setIsFormValidityFalseAction,
} from '../../store/errors/errors.actions';
import './Form.css';

type FormProps = {
  name: string;
  onSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
  children: React.ReactNode;
};

const Form: FC<FormProps> = ({ name, onSubmit, children }) => {
  const formRef = useRef<HTMLFormElement>(null);
  const dispatch = useDispatch();

  const onFormUpdate = (data: boolean) => {
    data
      ? dispatch(setIsFormValidityTrueAction())
      : dispatch(setIsFormValidityFalseAction());
  };

  const handleFormValidity = () => {
    const formValidityBoolean = formRef.current?.checkValidity();
    if (formValidityBoolean) {
      onFormUpdate(formValidityBoolean);
    }
  };

  return (
    <form
      action="#"
      className="popup__form"
      name={`myForm${name}`}
      onSubmit={onSubmit}
      ref={formRef}
      onChange={handleFormValidity}
    >
      {children}
    </form>
  );
};

export default Form;
