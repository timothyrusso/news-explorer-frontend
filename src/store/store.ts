import { legacy_createStore as createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer } from './root-reducer';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';
import { RootState } from './RootState';

const store = createStore(rootReducer, applyMiddleware(thunk));

export type AppDispatch = ThunkDispatch<RootState, unknown, AnyAction>;

export default store;
