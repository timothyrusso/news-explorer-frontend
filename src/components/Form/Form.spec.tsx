import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../store/store';
import Form from './Form';

describe('Footer component', () => {
  it('Should render the Footer component', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Form name={'myForm'} onSubmit={jest.fn()}>
          <input type="text" placeholder="Mock Input" />
          <button type="submit">Submit</button>
        </Form>
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
