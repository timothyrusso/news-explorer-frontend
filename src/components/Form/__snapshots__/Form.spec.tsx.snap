// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component Should render the Footer component 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="popup__form"
    name="myFormmyForm"
  >
    <input
      placeholder="Mock Input"
      type="text"
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
